lib/index.js

```diff
@@ -191,58 +241,48 @@ exports.set = function(path, val, o, special, map, _copying) {
 
   map || (map = K);
 
-  var parts = 'string' == typeof path
-    ? stringToParts(path)
-    : path;
-
-  if (!Array.isArray(parts)) {
-    throw new TypeError('Invalid `path`. Must be either string or array');
-  }
-
-  if (null == o) return;
-
-  for (var i = 0; i < parts.length; ++i) {
-    if (typeof parts[i] !== 'string' && typeof parts[i] !== 'number') {
-      throw new TypeError('Each segment of path to `set()` must be a string or number, got ' + typeof parts[i]);
-    }
-    // Silently ignore any updates to `__proto__`, these are potentially
-    // dangerous if using mpath with unsanitized data.
-    if (ignoreProperties.indexOf(parts[i]) !== -1) {
-      return;
-    }
-  }
-
-  // the existance of $ in a path tells us if the user desires
+  // the existence of $ in a path tells us if the user desires
   // the copying of an array instead of setting each value of
   // the array to the one by one to matching positions of the
   // current array. Unless the user explicitly opted out by passing
   // false, see Automattic/mongoose#6273
-  var copy = _copying || (/\$/.test(path) && _copying !== false),
+  let copy = _copying || (_copying !== false && path.indexOf('$') !== -1),
       obj = o,
-      part;
-
-  for (var i = 0, len = parts.length - 1; i < len; ++i) {
+      part,
+      i = 0,
+      len = parts.length - 1,
+      last = parts.length - 2,
+      j = 0,
+      jl = 0;
+
+  for (i = 0; i < len; ++i) {
     part = parts[i];
 
-    if ('$' == part) {
-      if (i == len - 1) {
+    // Silently ignore any updates to `__proto__`, these are potentially
+    // dangerous if using mpath with unsanitized data.
+    if (ignoreProperties.has(part)) {
+      return;
+    }
+
+    if (part === '$') {
+      if (i === last) {
         break;
       } else {
         continue;
       }
     }
 
-    if (Array.isArray(obj) && !/^\d+$/.test(part)) {
-      var paths = parts.slice(i);
+    if (Array.isArray(obj) && !numbers.has(part)) {
```

Comment:
Unfortunately I don't think this will work, because what happens if you try to set `arr.11`? `numbers` only contains digits.

Reply:
added unit test and fixed it